### ISSUE TITLE
[rigs] Add a network rig

### ISSUE THREAD

**Issue by TurboTurtle** [TurboTurtle/rig#3]
Placeholder to create a rig that monitors network activity. Initial idea is that users will provide a tshark packet filter string and once we get a single packet matching the filter, the rig will trigger.

**Comment by juanmasg** [comment]
Probably the most universal python module to achieve this would be Scapy. 
It supports bpf filters out of the box, and also gives access to all fields in all layers without the need to do extra string parsing of the output

**Comment by TurboTurtle** [comment]
Yeah, that's what I found as well, however the problem is that Scapy does not ship in RHEL currently. I opened an RFE a while back, but there has been no movement on it at all.

**Comment by juanmasg** [comment]
Well, it was a little bit heavyweight anyway :)

Getting rid of tcpdump will require at some point to rely on a wrapper for libpcap. There are quite a few out there, not sure if any of them is planned to be included in RHEL out of the box.

Worst case scenario, maybe one of these wrappers with the appropiate license can be just included as part of the rig package?

**Comment by juanmasg** [comment]
Reviving this issue :) I recently learned how to do this using just plain raw sockets. I'll try to provide a working prototype

**Comment by juanmasg** [comment]
@TurboTurtle as initial approach I'll try to provide command line options to define some basic matching options as IP address, port or protocol. Something like `--ip 10.0.0.1` or `--srcport 443/tcp`.

Allowing full user friendly bpf strings from the user and getting them translated into BPF JIT assembly without any 3rd party dependencies will require a lot of work and tunning.

See for example what a simple filter to match packets originating from a web service would translate to:

```
% sudo tcpdump -i any -n -d 'tcp and src port 80 or src port 443'
(000) ldh      [14]
(001) jeq      #0x86dd          jt 2	jf 11
(002) ldb      [22]
(003) jeq      #0x6             jt 4	jf 6
(004) ldh      [56]
(005) jeq      #0x50            jt 27	jf 19
(006) jeq      #0x2c            jt 28	jf 7
(007) jeq      #0x84            jt 9	jf 8
(008) jeq      #0x11            jt 9	jf 28
(009) ldh      [56]
(010) jeq      #0x1bb           jt 27	jf 28
(011) jeq      #0x800           jt 12	jf 28
(012) ldb      [25]
(013) jeq      #0x6             jt 14	jf 20
(014) ldh      [22]
(015) jset     #0x1fff          jt 28	jf 16
(016) ldxb     4*([16]&0xf)
(017) ldh      [x + 16]
(018) jeq      #0x50            jt 27	jf 19
(019) jeq      #0x1bb           jt 27	jf 28
(020) jeq      #0x84            jt 22	jf 21
(021) jeq      #0x11            jt 22	jf 28
(022) ldh      [22]
(023) jset     #0x1fff          jt 28	jf 24
(024) ldxb     4*([16]&0xf)
(025) ldh      [x + 16]
(026) jeq      #0x1bb           jt 27	jf 28
(027) ret      #262144
(028) ret      #0
```

The module would need to do the whole expression evaluation + translation to BPF asm in order to provide in-kernel filtering.

Another option would be to do just the expression evaluation and then doing the filtering in python after parsing at least layers 2, 3 and 4, which would not be even closely as efficient as compiling for the BPF JIT.

Do you have any commonly used filters, or are there any basic requirements that you have in mind? Would options `--srcip`, `--dstip`, `--protocol`, `--srcport`, `--dstport` do the job for now?

**Comment by TurboTurtle** [comment]
I honestly haven't looked at this deeply in a while, mainly due to the lack of movement of getting an existing module to support it available. If we can do this with raw sockets ourselves, that's definitely interesting!

Basic requirements for an MVP here I think would be a simple "trigger when we get X type of packet on Y address/port (optionally from Z source)". So, yes, the options you've listed look to be perfect for that use case.

To be a little more specific in the type of use case I envisioned for this, "trigger when we get a (or X number of) RST packet(s) from srcip 10.11.12.13 on dstip 10.9.8.7"

**Comment by TurboTurtle** [comment]
I like that list. I had some initial hesitance over `packet`, but it makes more sense the more I think on it, and I continue to not be able to come up with a better alternative anyways :wink:
